docs/source/concepts/interoperability.rst

```diff
@@ -0,0 +1,237 @@
+The flat transaction stream by `P2` shows omits the non-consuming exercise choices.
+It nevertheless contains the three view change events ``EnterLocalLedgerEvent``, ``LeaveLocalLedgerEvent``, and ``EnterLocalLedgerEvent`` before the ``ArchivedEvent``.
+This is because the endpoint cannot know at the ``LeaveLocalLedgerEvent`` that there will be another ``EnterLocalLedgerEvent``.
+
+In contrast, `P1` need not output the ``EnterLocalLedgerEvent``\ s and ``LeaveLocalLedgerEvent``\ s at all in this example because `P1` hosts `A` on both ledgers.
+
+
+Local Ledgers with Enter and Leave Actions
+==========================================
+
+The ``EnterLocalLedgerEvent`` and ``LeaveLocalLedgerEvent`` events are included in the transactions of the local ledger that the participant maintains for a party.
+They are abbreviated as **Enter** and **Leave** actions.
+**Enter** counts as a contract activation and **Leave** as a contract deactivation.
+The :ref:`local ledger structure <local-ledger-structure>` for interoperable ledgers generalizes as follows.
+
+The transactions in the local ledger contain all the actions of the DAML Ledger Model and additionally the ``EnterLocalLedgerEvent`` and ``LeaveLocalLedgerEvent`` events,
+which are referred to as **Enter** and **Leave** actions.
+As before, the stream of transactions and transaction trees are derived as a topological sort of the local ledger for the set of parties.
+
+It suffices to extend :ref:`definition of activation and deactivation <local-ledger-structure>` actions as follows.
+The actual definition of local ledger remains the same except that every local ledger implicitly identifies the set of DAML Ledgers it aggregates.
+
+Definition »activation, deactivation«
+
+  * A **Enter** action on a contract **activates the contract**.
+
+  * A **Leave** action on a contract **deactivates the contract**.
+
+
+.. _interoperable-causality:
+
+Causality for Interoperable Ledgers
+***********************************
+
+The :ref:`ordering guarantees <order-consistency>` between the local ledger and the transaction service does not need to be changed for interoperability.
+The order consistency of several local ledgers and the shared ledger, however, need to be generalized for interoperability.
+
+The virtual shared ledger
+=========================
+
+While every DAML Ledger may keep a physical copy of its shared ledger,
+there is no place that records the result of the interoperability protocol between those ledgers;
+it merely ensures that the individual shared ledgers are consistent.
+Like for local ledgers, consistency is formalized by the existence of a *virtual* shared ledger.
+
+Definition »virtual shared ledger«
+  A **virtual shared ledger** for a set `X` of interoperable DAML Ledgers
+  is a finite labelled directed acyclic graph of transactions that satisfies the following conditions:
+
+  * It satisfies the conditions for a shared ledger.
+
+  * It does not contain **Enter** nor **Leave** actions.
```

Comment:
Yes, it would make the consistency definitions easier, but `Enter` and `Leave` do not really make sense on the virtual shared ledger. `Enter` and `Leave` are specific to the local ledgers for a particular party at a particular participant node. Moreover, if we include them in the virtual shared ledger in some way, we'd either have to include such actions for each participant-party pair individually (i.e., label them accordingly) or they would induce further causality requirements between participants. But I don't see why we should promise consistency for Enter/Leave. In my view, they are just artefacts of the limited visibility of participants.

Moreover, if we include them in the virtual shared ledger, then we have to get rid of them in the step to the DAML ledger model's ledger.